crates/sui-core/src/consensus_adapter.rs

```diff
@@ -312,28 +306,49 @@ impl ConsensusAdapter {
         } else {
             None
         };
-        self.authority
-            .database
-            .epoch_store()
-            .insert_pending_consensus_transactions(&transaction)?;
+        epoch_store.insert_pending_consensus_transactions(&transaction)?;
         if let ConsensusTransactionKind::UserTransaction(certificate) = &transaction.kind {
             self.pending_certificates
                 .lock()
                 .insert(*certificate.digest());
         }
-        Ok(self.submit_unchecked(transaction))
+        Ok(self.submit_unchecked(transaction, epoch_store.clone()))
     }
 
-    fn submit_unchecked(self: &Arc<Self>, transaction: ConsensusTransaction) -> JoinHandle<()> {
+    fn submit_unchecked(
+        self: &Arc<Self>,
+        transaction: ConsensusTransaction,
+        epoch_store: Arc<AuthorityPerEpochStore<AuthoritySignInfo>>,
+    ) -> JoinHandle<()> {
         // Reconfiguration lock is dropped when pending_consensus_transactions is persisted, before it is handled by consensus
-        let async_stage = self.clone().submit_and_wait(transaction);
+        let async_stage = self.clone().submit_and_wait(transaction, epoch_store);
         // Number of this tasks is limited by `sequencing_certificate_inflight` limit
         let join_handle = spawn_monitored_task!(async_stage);
         join_handle
     }
 
+    async fn submit_and_wait(
+        self: Arc<Self>,
+        transaction: ConsensusTransaction,
+        epoch_store: Arc<AuthorityPerEpochStore<AuthoritySignInfo>>,
+    ) {
+        let epoch_terminated = epoch_store.wait_epoch_terminated().boxed();
+        let submit_and_wait = self
+            .submit_and_wait_inner(transaction, &epoch_store)
+            .boxed();
+        // When epoch_terminated signal is received all pending submit_and_wait_inner are dropped.
+        //
+        // This is needed because submit_and_wait_inner waits on read_notify for consensus message to be processed,
+        // which may never happen on epoch boundary.
+        select(submit_and_wait, epoch_terminated).await;
+    }
+
     #[allow(clippy::option_map_unit_fn)]
-    async fn submit_and_wait(self: Arc<Self>, transaction: ConsensusTransaction) {
+    async fn submit_and_wait_inner(
+        self: Arc<Self>,
+        transaction: ConsensusTransaction,
+        epoch_store: &Arc<AuthorityPerEpochStore<AuthoritySignInfo>>,
```

Comment:
Is the idea of passing epoch_store around to make sure we always access the same epoch store? Given that we will eventually need to take a lock of the state and pass it down, this is unnecessary. More importantly, it's not effective: for instance, inside this function, the call to `consensus_message_processed_notify` will actually access epoch store of the state.

Reply:
Good point about `consensus_message_processed_notify`, looks like it needs to be moved to PerEpochStore.

In terms of taking lock and passing it down - the way I see it, this exactly will mean that we will be passing `RwLockReadGuard<AuthorityPerEpochStore>`, instead of `Arc<AuthorityPerEpochStore>`, so this change moves code in right direction

Reply:
Moved `consensus_message_processed_notify` into PerEpochStore

Reply:
This is problem out of the scope of this PR, but just to discuss it before I forget. I see two potential issues of locking AuthorityPerEpochStore instead of locking AuthorityState:
1. We could have code patterns where we are accessing epoch store and non-epoch store in the same call path, and we could be finishing the non-epoch store writes abut then fails the epoch-store writes, leading to somewhat inconsistent data (not sure if we have this, but a concern)
2. We also have a committee field in AuthorityState, that needs to be somewhat in-sync with the per-epoch store. So if we are only locking the per-epoch store, we may need to move that committee field into the per-epoch store.

Reply:
1. I don't have super good answer right now. For some components like sending from ConsensusAdapter, it seems that we can live with only using PerEpochStore. For some components like TransactionManager that truly do need to have both PerEpoch and permanent store(as it works with effects and objects) we'll need to figure something out.

2. Yes, I actually do think that we need to move committee to per-epoch store. We already pass committee there and we can just store it in local field. I am not sure however how hard is it to remove committee from Authority. But intuitively, this seem to be a move in the right direction.